// lavalink/dist/base/Cluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const ClusterNode_1 = require("../ClusterNode");
class Cluster extends events_1.EventEmitter {
    constructor(options) {
        super();
        this.nodes = [];
        if (options)
            this.spawn(options);
    }
    spawn(options) {
        if (Array.isArray(options))
            return options.map(opt => this.spawn(opt));
        const node = new ClusterNode_1.default(this, options);
        this.nodes.push(node);
        return node;
    }
    sort() {
        // filter nodes for open ws connections and restrict to specified tag (if provided)
        return this.nodes.slice().sort((a, b) => {
            if (!a.stats || !b.stats)
                return -1;
            return (a.stats.cpu ? a.stats.cpu.systemLoad / a.stats.cpu.cores : 0)
                - (b.stats.cpu ? b.stats.cpu.systemLoad / b.stats.cpu.cores : 0);
        });
    }
    has(guildID) {
        return this.nodes.some(node => node.players.has(guildID));
    }
    get(guildID) {
        let node = this.nodes.find(node => node.players.has(guildID));
        if (!node)
            node = this.sort().find(node => this.filter(node, guildID));
        if (node)
            return node.players.get(guildID);
        throw new Error('unable to find appropriate node; please check your filter');
    }
    voiceStateUpdate(state) {
        return this.get(state.guild_id).node.voiceStateUpdate(state);
    }
    voiceServerUpdate(server) {
        return this.get(server.guild_id).node.voiceServerUpdate(server);
    }
}
exports.default = Cluster;
